core/shared/src/main/scala/org/http4s/multipart/PartReceiver.scala

```diff
@@ -0,0 +1,206 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.multipart
+
+import cats.Applicative
+import cats.ApplicativeError
+import cats.MonadThrow
+import cats.effect.kernel.Resource
+import fs2.Compiler
+import fs2.Pipe
+import fs2.Pull
+import fs2.Pure
+import fs2.RaiseThrowable
+import fs2.Stream
+import fs2.io.file.Files
+import fs2.io.file.Flags
+import fs2.io.file.Path
+import org.http4s.DecodeFailure
+import org.http4s.EntityDecoder
+import org.http4s.Headers
+import org.http4s.InvalidMessageBodyFailure
+
+/** Represents the decoding process of a single "part" in a `multipart/form-data` message.
+  *
+  * Used in conjunction with `MultipartReceiver` and `MultipartDecoder.fromReceiver`
+  * to create robust `EntityDecoder`s with fail-fast behavior.
+  */
+trait PartReceiver[F[_], A] {
+  def receive(part: Part[F]): Resource[F, Either[DecodeFailure, A]]
+
+  def map[B](f: A => B): PartReceiver[F, B] =
+    part => this.receive(part).map(_.map(f))
+
+  def mapWithHeaders[B](f: (Headers, A) => B): PartReceiver[F, B] =
+    part => this.receive(part).map(_.map(f(part.headers, _)))
+
+  def tapStart(onStart: F[Unit]): PartReceiver[F, A] =
+    part => Resource.eval(onStart).flatMap(_ => this.receive(part))
+
+  def tapResult(onResult: Either[DecodeFailure, A] => F[Unit]): PartReceiver[F, A] =
+    part => this.receive(part).evalTap(onResult)
+
+  def tapRelease(onRelease: F[Unit])(implicit F: Applicative[F]): PartReceiver[F, A] =
+    part => Resource.make(F.unit)(_ => onRelease).flatMap(_ => this.receive(part))
+
+  def preprocess(transformPartBody: Pipe[F, Byte, Byte]): PartReceiver[F, A] =
+    part => this.receive(part.copy(body = transformPartBody(part.body)))
+
+  def withSizeLimit(limit: Long)(implicit F: ApplicativeError[F, Throwable]): PartReceiver[F, A] =
+    preprocess(PartReceiver.limitPartSize[F](limit))
+
+  def rejectIfFile: PartReceiver[F, A] =
+    part =>
+      if (part.filename.isDefined) {
+        val effectivePartName = part.name.fold("<unnamed part")(name => s"'$name'")
+        Resource.pure(
+          Left(InvalidMessageBodyFailure(s"File data not allowed in $effectivePartName"))
+        )
+      } else {
+        this.receive(part)
+      }
+
+  def rejectIfNotFile: PartReceiver[F, A] =
+    part =>
+      if (part.filename.isEmpty) {
+        val effectivePartName = part.name.fold("<unnamed part")(name => s"'$name'")
+        Resource.pure(Left(InvalidMessageBodyFailure(s"File data required in $effectivePartName")))
+      } else {
+        this.receive(part)
+      }
+}
+
+object PartReceiver {
+
+  /** Creates a PartReceiver which decodes the part body to a String.
+    *
+    * The decoding will use UTF-8 unless the part provides a `Content-Type` header indicating otherwise.
+    */
+  def bodyText[F[_]](implicit F: RaiseThrowable[F], cmp: Compiler[F, F]): PartReceiver[F, String] =
```

Comment:
`F: RaiseThrowable[F], cmp: Compiler[F, F]` can be replaced by `F: Concurrent[F]`

Reply:
Doing so broke `MultipartReceiver.auto` (had to swap it from Sync to Concurrent) but I think that's fine so I went ahead with it.